### ISSUE TITLE
Copy ~/.cache/x into Docker

### ISSUE THREAD

**Issue by yeldarby** [replicate/cog#320]
Every time I do `cog predict`, the `huggingface` package is downloading the 500MB GPT2 pretrained weights into `~/.cache/huggingface` which is quite slow & bandwidth intensive.

I can't figure out how to bundle these into the Docker.

So far I've tried
* Using `run` in `cog.yaml` (doesn't work because it doesn't have access to the file yet)
* Copying `~/.cache/huggingface` into my working dir and having my `predict.py` copy the files at the top (which doesn't work for unknown reasons; possibly everything outside of `/src` is read-only) `os.system("cp -r /src/cache/huggingface /root/.cache/huggingface")`

**Comment by zeke** [comment]
Hi @yeldarby. Thanks for opening this issue and following up with the fix you found.

It looks like this issue is specific to the `huggingface` package, so I'm not sure what generic changes we could make here to prevent other users of Cog from running into this issue. Do you have any suggestions on how we might improve behavior or documentation around this in a generalizable way?

**Comment by yeldarby** [comment]
It's not just `huggingface`; other libraries like `openai/clip` and `torch` also store data in `~/.cache`.

Ideally make a way to mount volumes in custom locations where data will persist between runs. Or being able to add `COPY` lines to the `Dockerfile` to move files into the proper location on the container's filesystem.

**Comment by hangtwenty** [comment]
### I want this!

As a user I want something like this too. In Docker terms I'm leaning more towards `ENV` than `COPY` as a way to address it.

## Brainstorming

### Security note

**It's very important** that the host's `~/.cache` does not get automatically pulled in, because a user could have private/sensitive/unrelated data in there. (Aside from bloat.)

### `$XDG_CACHE_HOME`

`~/.cache` is a conventional default value for [`$XDG_CACHE_HOME`](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html) in Linux desktops. Some libraries _may_ hardcode `~/.cache` but many libraries respect `$XDG_CACHE_HOME`, including [PyTorch](https://pytorch.org/docs/stable/hub.html#:~:text=TORCH_HOME%20is%20set.-,%24XDG_CACHE_HOME,-/torch/hub%2C%20if), [huggingface](https://huggingface.co/transformers/v4.0.1/installation.html#caching-models), [pip](https://stackoverflow.com/a/64180835) ... (Aside from ML or Python, there are [many](https://wiki.archlinux.org/title/XDG_Base_Directory), [many](https://github.com/search?q=XDG_CACHE_HOME&type=code) [more](https://wiki.debian.org/DotFilesList) ...)

### `cog.yaml` could set a default `XDG_CACHE_HOME`

Maybe `cog` could be pre-loaded with some environment variable defaults. Inside the container, it could make sense to set `XDG_CACHE_HOME` to `/src/cog_cache` (or `/src/cache` or `/src/.cache` ... It needs to avoid name-collision, so `cache` might be too generic to use as the default.). 

Given that `/src` is the only directory retained, it makes sense to set `XDG_CACHE_HOME` to a subdirectory of `/src`. (Just like @yeldarby shared above, when setting it with `os.environ`).

It _might_ even make sense to make that a default behavior of `cog` (with `cog.yaml` able to override or disable it).

#### `TRANSFORMERS_CACHE`, `TORCH_HOME` ...

* **A barebones solution for `ENV` would go a long way.** 
    * Maybe `cog.yaml` could allow adding a list of environment variables, which `cog` will add as `ENV` directives in the generated Dockerfile.

* Maybe `cog` could start including some `ENV` "presets" for popular ML libraries.
    * Maybe there could be a subcommand like `cog init` but for _updating_ a `cog.yaml` to add automatically inferred defaults. (Name ideas: `cog assist`/`cog magic`/`cog revise` ...) 
    * It would read `cog.yaml`, then update it.
    * Example: detecting `torch` or `torchvision` in `python_packages` could add a default `TORCH_HOME` value. Detecting `huggingface` could set a default for `TRANSFORMERS_CACHE`. And so on.

<details>
<summary>What about on the host?</summary>

* As a user I'd only need to override these on the host if the fetching is happening on the host, i.e. when there's a script to download models/etc into the workdir, to be run manually before `cog build`/`cog predict`. (Or i.e. when there's an ad-hoc command to run according to a README etc.) 
* As a user I'd see some relevant env var defaults appear in `cog.yaml` after doing `cog assist` (or whatever it's named). 
    * For example, in the `cog.yaml` if I saw `TRANSFORMERS_HOME=/src/cog_cache  # huggingface cache` 
    * ... I think that'd suffice to remind me what I need to do on the host/ before doing ad-hoc fetches. (i.e. on the host I would be reminded to do `export TRANSFORMERS_HOME="$(pwd)/cog_cache"` before running the fetching script/commands.)

* After seeing where users get snagged, could add some more direct hints.

</details>

##### `--interactive`

* If cog added something like this `cog assist`/`cog revise` idea, could imagine it evolving to have an `--interactive` question/answer mode. It could ask for confirmation about ambiguous or less-common things.
* Could imagine `cog assist`/`cog revise` evolving over time, to include smarts beyond environment variables. For example detecting something-is-missing from `system_packages` based on patterns in `python_packages`. I think the `cog` team wants to add smarts like that over time; maybe "a subcommand that revises the `cog.yaml`" is a good place to put that functionality.
* An `--interactive` mode would allow that to grow without too much magic in the normal usage.

## TLDR

* `cog.yaml` should provide a way to add `ENV` directives
* It'd be awesome to have a `cog` subcommand like `cog assist`/`cog revise` that infers some things from `cog.yaml`, then updates `cog.yaml` in-place.

**Comment by hangtwenty** [comment]
Thought of a related idea.

##### "Untracked files" warning for `~/.cache` inside Docker

* It's helpful when `git` or your IDE warns you about untracked files you may have meant to commit.
* Maybe `cog` could try something that for `$XDG_CACHE_HOME`, `$HOME/.cache`, etc. As library-specific cases are added, those could be checked too. Maybe this would be a flag like `--inspect-cache`, on `cog build`|`run`|`predict`. It could compare the filesystem data before/after it runs a subcommand, and do some diffing. When it appears that the cache has grown (-- significantly? --) in size, it would print info about what changed/ give hints on what might want to be cached.
    * Even fancier: it could be _watching_ the filesystem as things run, and it could print the `--inspect-cache` warnings as they happen. **I'm curious if that would be useful in your case, @yeldarby**: if `cog` could "notice" that there's a big download happening, and print/yell about it as it's happening... Would that be helpful?

**Comment by hangtwenty** [comment]
(Re-opened the PR to add support for `ENV`, and a default `XDG_CACHE_HOME`.)

**Comment by markphelps** [comment]
This issue has been inactive for over 4 years, so we're closing it as stale. The discussion evolved into a feature request for `ENV` support in `cog.yaml` and default `XDG_CACHE_HOME` handling. There doesn't appear to be a clear resolution path on this thread. If this is still a need with the current version of Cog, please open a new issue and reference #320.
